automata/fa/fa.py

```diff
@@ -2,11 +2,245 @@
 """Classes and methods for working with all finite automata."""
 
 import abc
+import itertools
+import os
+import pathlib
+import typing
+import uuid
+from collections import defaultdict
+
+import graphviz
+from coloraide import Color
 
 from automata.base.automaton import Automaton
 
 
 class FA(Automaton, metaclass=abc.ABCMeta):
     """An abstract base class for finite automata."""
 
-    pass
+    @staticmethod
+    def get_state_name(state_data):
+        """
+        Get an string representation of a state. This is used for displaying and
+        uses `str` for any unsupported python data types.
+        """
+        if isinstance(state_data, str):
+            if state_data == "":
+                return "λ"
+
+            return state_data
+
+        if isinstance(state_data, (set, frozenset, list, tuple)):
+            inner = ", ".join(FA.get_state_name(sub_data) for sub_data in state_data)
+            if isinstance(state_data, (set, frozenset)):
+                return '{' + inner + '}'
+
+            if isinstance(state_data, tuple):
+                return '(' + inner + ')'
+
+            if isinstance(state_data, list):
+                return '[' + inner + ']'
+
+        return str(state_data)
+
+    @abc.abstractmethod
+    def iter_states(self):
+        """Iterate over all states in the automaton."""
+
+    @abc.abstractmethod
+    def iter_transitions(self):
+        """
+        Iterate over all transitions in the automaton. Each transition is a tuple
+        of the form (from_state, to_state, symbol)
+        """
+
+    @abc.abstractmethod
+    def is_accepting(self, state):
+        """Check if a state is an accepting state."""
+
+    @abc.abstractmethod
+    def is_initial(self, state):
+        """Check if a state is an initial state."""
+
+    def show(
+        self,
+        input_str: str | None = None,
+        save_path: str | os.PathLike | None = None,
```

Comment:
@caleb531 somewhat tangentially related, but how would you feel about trying to add type annotations to the library again? I know this got dropped a while ago, but I'm realizing it makes writing code easier, since editors like VSCode use these annotations for autocomplete features. The annotations don't have to be super detailed, just simple things like "-> DFA" or "-> Set" and things like that.

Reply:
@eliotwrobson You know, I was actually thinking about this the other day when I noticed type annotations in this PR lol.

I recall one of the places we left off (in #43) was [an issue you were having](https://github.com/caleb531/automata/pull/43#issuecomment-1101880464) around changing function signatures in derived classes, and mypy not being happy with that. But perhaps the tooling is better now (especially with a different type checker like Pylance) to where that won't be much of an issue.

But yes, I like the idea of trying to integrate types with this library, especially since it _is_ a library and not just application code. And if we have a reasonable solution to the above issue, then I think we'll be golden.

Is this something you would be up for taking on again?

Reply:
> I recall one of the places we left off (in #43) was [an issue you were having](https://github.com/caleb531/automata/pull/43#issuecomment-1101880464) around changing function signatures in derived classes, and mypy not being happy with that. But perhaps the tooling is better now (especially with a different type checker like Pylance) to where that won't be much of an issue.

I think that Pyright (the typechecker used by Pylance) will also complain about this (it violates Liskov), but this might be as simple as using a wrapper function. The main issue was adding annotations that had issues with 3.6, but that's no longer an issue.

> Is this something you would be up for taking on again?

Sure! I can take a look and start playing around.